app/models/form.rb

```diff
@@ -175,9 +175,10 @@ def visible_questionings
     questionings.reject{|q| q.hidden?}
   end
 
-  # returns questionings that work with sms forms and are not hidden
+  # returns hash of questionings that work with sms forms and are not hidden
   def smsable_questionings
-    questionings.reject{|q| q.hidden? || !q.question.qtype.smsable?}
+    smsable_questionings = questionings.reject{|q| q.hidden? || !q.question.qtype.smsable?}
+    smsable_questionings.index_by.with_index { |_, i| i + 1 }
```

Comment:
Yeah, I think part of the confusion might be that the C code shown in ruby's documentation site is severely clipped and more or less just farms out the work to methods defined elsewhere. I struggle reading C so it's still a mystery to me looking at the full source, but maybe you'll find https://github.com/ruby/ruby/blob/ruby_2_2/enumerator.c#L495 more interesting.

I'm fine with changing it to the map with index though, I have a preference for clear code over clever code.